cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricsForWindow.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.cruisecontrol.monitor.sampling.aggregator;
+
+import com.linkedin.cruisecontrol.common.LongGenerationed;
+import com.linkedin.cruisecontrol.monitor.sampling.MetricSample;
+import com.linkedin.cruisecontrol.monitor.sampling.Snapshot;
+import com.linkedin.cruisecontrol.resource.Resource;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.cruisecontrol.monitor.sampling.aggregator.SnapshotAndImputation.Imputation.AVG_ADJACENT;
+import static com.linkedin.cruisecontrol.monitor.sampling.aggregator.SnapshotAndImputation.Imputation.AVG_AVAILABLE;
+import static com.linkedin.cruisecontrol.monitor.sampling.aggregator.SnapshotAndImputation.Imputation.FORCED_INSUFFICIENT;
+import static com.linkedin.cruisecontrol.monitor.sampling.aggregator.SnapshotAndImputation.Imputation.NONE;
+import static com.linkedin.cruisecontrol.monitor.sampling.aggregator.SnapshotAndImputation.Imputation.NO_VALID_IMPUTATION;
+
+
+/**
+ * A class hosting all the metric samples for certain snapshot window.
+ */
+public class MetricsForWindow<E> extends LongGenerationed {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsForWindow.class);
+  private final Map<E, E> _identityEntityMap;
+  private final long _window;
+  private final Map<E, AggregatedMetrics> _entityMetrics;
+  private final AtomicLong _numSamples;
+  private volatile WindowAggregationResult<E> _cachedAggregationResult;
+  private volatile long _prevWindowGenerationForCache;
+  private volatile long _nextWindowGenerationForCache;
+
+  MetricsForWindow(Map<E, E> identityEntityMap, long window) {
+    super(-1);
+    _identityEntityMap = identityEntityMap;
+    _window = window;
+    _entityMetrics = new ConcurrentHashMap<>();
+    _numSamples = new AtomicLong(0);
+    _cachedAggregationResult = null;
+    _prevWindowGenerationForCache = -1L;
+    _nextWindowGenerationForCache = -1L;
+  }
+
+  long window() {
+    return _window;
+  }
+
+  long numSamples() {
+    return _numSamples.get();
+  }
+
+  Map<E, AggregatedMetrics> entityMetrics() {
+    return Collections.unmodifiableMap(_entityMetrics);
+  }
+
+  /**
+   * Add a sample to this window
+   * @param sample the sample to add.
+   */
+  void addSample(MetricSample<E> sample, Long generation) {
+    AggregatedMetrics aggMetrics =
```

Comment:
Sorry if this does not apply.  I've forgotten the threading model that cruise control is using.  _identityEntityMap is just Map<> and so this might be multithread safe.

Reply:
The implementation is a ConcurrentHashMap, so it should be safe. I can change the declaration to be ConcurrentMap.